akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSpec.scala

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import scala.concurrent.{ Await, ExecutionContext, Future, Promise }
+import scala.concurrent.duration._
+
+import java.util.concurrent.CompletionStage
+import scala.compat.java8.FutureConverters._
+
+import akka.NotUsed
+import akka.stream._
+import akka.stream.stage.{ GraphStage, GraphStageLogic, OutHandler }
+import akka.stream.scaladsl._
+import akka.stream.impl.fusing.GraphStages.FutureFlattenSource
+
+import org.scalatest.concurrent.PatienceConfiguration.Timeout
+
+import akka.testkit.EventFilter
+import akka.stream.testkit.{ StreamSpec, TestSubscriber }
+import akka.stream.testkit.Utils.assertAllStagesStopped
+
+class FutureFlattenSpec extends StreamSpec {
+  val materializer = ActorMaterializer()
+
+  "Future source" must {
+    {
+      implicit def m = materializer
+      implicit def ec = materializer.executionContext
+
+      "flatten elements" in assertAllStagesStopped {
+        val subSource: Source[Int, String] =
+          Source(List(1, 2, 3)).mapMaterializedValue(_ ⇒ "foo")
+
+        val futureSource = new FutureFlattenSource(Future(subSource))
+        val source: Source[Int, Future[String]] = Source.fromGraph(futureSource)
+
+        val materialized = Promise[String]()
+        val watched: Source[Int, NotUsed] = source.watchTermination() { (m, d) ⇒
+          materialized.completeWith(d.flatMap(_ ⇒ m))
+          NotUsed
+        }
+
+        val p = watched.runWith(Sink asPublisher false)
+        val c = TestSubscriber.manualProbe[Int]()
+        p.subscribe(c)
+
+        val sub = c.expectSubscription()
+        sub.request(5)
+
+        c.expectNext(1)
+        c.expectNext(2)
+        c.expectNext(3)
+
+        c.expectComplete()
+
+        materialized.future.futureValue(Timeout(3.seconds)) should ===("foo")
+      }
+
+      "flatten elements from a completion stage" in assertAllStagesStopped {
+        val subSource: Graph[SourceShape[Int], Int] =
+          Source(List(1, 2, 3)).mapMaterializedValue(_ ⇒ 1)
+
+        val future = Future(subSource)
+        val stage: CompletionStage[Graph[SourceShape[Int], Int]] = future.toJava
+        val g = Source.fromSourceCompletionStage(stage)
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.toScala.futureValue should ===(1)
+        fut.futureValue should ===(List(1, 2, 3))
+      }
+    }
+
+    "be cancelled before the underlying Future completes" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
```

Comment:
`    "be cancelled before the underlying Future completes" in assertAllStagesStopped {` is not too long for scalariform or compared to what we normally stay around in the Akka sources